### PR TITLE
Add unit tests for the functions in utils.py

### LINKED ISSUE #6: Add unit tests for the functions in utils.py

### PR DESCRIPTION
**Pull request by Shashi456**
Fixes #6

These are just very basic unit tests , which is my understanding of them. Since we are just checking each module individually and seeing if they do the job they are supposed to do. Also another thing i wanted to ask is how could i go about writing tests for functions which had no parameters, should i just check the existence of what has been returned? Thanks for your help.

@marco-c I will clean up the code and use tempfile in the upcoming commit, this was just for a basic check to see if you were expecting similar unit tests.

### CODE REVIEW

File: tests/test_utils.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
+import os
+import csv
+import keras
+from keras.preprocessing.image import load_img, img_to_array
+
+
+def mkdir(dir_name):
```
Comment (marco-c) [inline tests/test_utils.py]: The functions that need to be tested should not be here.

**Comment by marco-c** [comment]
>  Also another thing i wanted to ask is how could i go about writing tests for functions which had no parameters, should i just check the existence of what has been returned?

For functions that require the `data` directory to exist, don't write any test.
For other functions with no parameters (e.g. `get_bugs`), just make some basic assertion on the result.

**Comment by marco-c** [comment]
> These are just very basic unit tests , which is my understanding of them. Since we are just checking each module individually and seeing if they do the job they are supposed to do.

Yes, this is what we should do.

**Comment by marco-c** [comment]
> @marco-c I will clean up the code and use tempfile in the upcoming commit, this was just for a basic check to see if you were expecting similar unit tests.

OK, let me know when the PR is ready and I'll review it.

**Comment by Shashi456** [comment]
@marco-c So i've written basic tests for 6 functions, just asserting the existence of the objects in case of functions with no specific parameters. For writing a couples_generator, getImageGenerator there is image_test or image_train so i have written no function for the same.

File: tests/test_utils.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
+import os
+from autowebcompat import utils
+
+
+def test_mkdir():
+    direc_path = os.path.dirname(os.path.realpath(__file__)) + "test"
+    utils.mkdir(direc_path)
+    assert(os.path.exists(direc_path))
```
Comment (marco-c) [inline tests/test_utils.py]: You can assert that this is a directory.
Also, you can test calling it again to check that it doesn't fail when the directory already exists.

File: tests/test_utils.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+
+label = {}
```
Comment (marco-c) [inline tests/test_utils.py]: Not sure you need to define this globally.

File: tests/test_utils.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+
+def test_read_labels():
+    labels = utils.read_labels(file_name='labels.csv')
+    assert('labels' in locals() or 'labels' in globals())
```
Comment (marco-c) [inline tests/test_utils.py]: This assertion is strange. Just assert that labels is not null and that it is a dictionary.

File: tests/test_utils.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+
+def test_write_labels():
+    utils.write_labels(label, file_name='test.csv')
+    fname = os.path.dirname(os.path.realpath(__file__)) + "\\test.csv"
+    assert(os.path.exists(fname))
```
Comment (marco-c) [inline tests/test_utils.py]: Also assert that you can use read_labels to read the labels back.

File: tests/test_utils.py [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+
+images = {"Image.png": [23434, 324124]}
```
Comment (marco-c) [inline tests/test_utils.py]: This is unused.

File: tests/test_utils.py [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+
+def test_load_image():
+    img = utils.load_image("Image.png")
+    assert(img)
```
Comment (marco-c) [inline tests/test_utils.py]: assert that this is a numpy 2x2 array

File: tests/test_utils.py [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+
+def test_get_bugs():
+    Object = utils.get_bugs()
+    assert(Object)
```
Comment (marco-c) [inline tests/test_utils.py]: assert that this is a list

File: tests/test_utils.py [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+
+
+def test_balance():
+    Sample = [[1, 0, 0], [0, 1, 0], [0, 0, 1]]
+    it = utils.balance(Sample)
+    assert(it)
```
Comment (marco-c) [inline tests/test_utils.py]: You can actually assert what balance returns here.

**Comment by marco-c** [comment]
> For writing a couples_generator, getImageGenerator there is image_test or image_train so i have written no function for the same.

I'm not sure what you mean here, can you expand?

**Comment by marco-c** [comment]
Nit: Can you put the tests in the same order as the functions of test_utils.py?

**Comment by marco-c** [comment]
Also, don't forget to use tempfile to create the temporary directory where you store the temporary test files.

File: tests/test_utils.py (hunk 4 above)
Comment (Shashi456) [reply tests/test_utils.py]: i was just asserting the existence of the variable

Comment (marco-c) [reply tests/test_utils.py]: You have defined it at line 15, so it clearly exists 😄

**Comment by Shashi456** [comment]
@marco-c  I have added temporary directory to the unit tests wherever I am making a directory or a file. I have also made other minor changes you asked me to do. I had two doubts though, What is balance returning? and I was getting a row index out of range error while reading the written labels do you have any idea what could cause it?

File: tests/.pytest_cache/v/cache/lastfailed [hunk 10]
```diff
@@ -0,0 +1,3 @@
+{
+  "test_utils.py::test_answer": true
```
[the remark this answers is not in the document]
Comment (marco-c) [reply tests/.pytest_cache/v/cache/lastfailed]: Indeed, and it should also be removed from the PR.

File: tests/test_utils.py [hunk 11]
```diff
@@ -0,0 +1,40 @@
+import os
+import numpy as np
+from tempfile import TemporaryDirectory
+from autowebcompat import utils
+
+
+def test_get_bugs():
+    Object = utils.get_bugs()
```
Comment (marco-c) [inline tests/test_utils.py]: Nit: rename `Object` to `bugs`.

File: tests/test_utils.py [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+    assert(isinstance(Object, list))
+
+
+def test_mkdir():
+    d = TemporaryDirectory()
+    direc_path = os.path.join(d.name + "\\test")
+    utils.mkdir(direc_path)
+    assert(os.path.isdir(direc_path))
```
Comment (marco-c) [inline tests/test_utils.py]: Second part from https://github.com/marco-c/autowebcompat/pull/63#discussion_r168962349

File: tests/test_utils.py [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+
+
+def test_load_image():
+    img = utils.load_image("Image")
```
Comment (marco-c) [inline tests/test_utils.py]: This file doesn't exist in the repository, so I'm not sure the test will pass.

**Comment by marco-c** [comment]
> What is balance returning?

You can see the function implementation or the callers of the functions to understand. If you can't figure it out, please remove the current `test_balance` and file a follow-up issue to add a test for `balance`.

> and I was getting a row index out of range error while reading the written labels do you have any idea what could cause it?

I'm not sure, as I haven't seen what you were trying to do. If you can't figure it out, please file a follow-up issue to add the test.

**Comment by marco-c** [comment]
Please also update the .travis.yml file to actually run the tests.

**Comment by marco-c** [comment]
I've noticed there's a nice pytest feature that you can use instead of manually using `tempfile`: https://docs.pytest.org/en/latest/tmpdir.html.

File: tests/test_utils.py (hunk 13 above)
Comment (Shashi456) [reply tests/test_utils.py]: @marco-c  earlier i had the load_image function in the file and i was using a random dictionary to just check if the code was working, Because an empty file name would result in a parameter not declared error. Should i remove the test ?

File: tests/test_utils.py (hunk 12 above)
Comment (Shashi456) [reply tests/test_utils.py]: @marco-c  Instead of writing a test for it , would it not make more sense to change the mkdir function to check if a directory with the same name exists and not make it if it does? since the function the doesnt return any error message it is pretty hard to assert on anything . If some error message were returned i could assert on the same .

Comment (marco-c) [reply tests/test_utils.py]: You don't need to add an assertion, the assertion is implied in the fact that `utils.mkdir` doesn't throw an exception.

File: tests/test_utils.py (hunk 13 above)
Comment (marco-c) [reply tests/test_utils.py]: You can add a `parent_dir` argument to `load_image`, with the default being `data_resized`.
Then in the test you can create a fake image in the temporary directory, pass the temporary directory as parent_dir to load_image and then assert that the image is loaded as a 2x2 numpy array with the correct dimensions.

**Comment by Shashi456** [comment]
@marco-c  i've made all the minor changes you asked me to, i have removed the balance test and will open a pr for the same, because balance seems to be interacting with Couples-generator but we aren't writing tests for functions which need the data directory. One last thing is the error in the Travis-cl build it is giving an image not found error with the Load_image test, while I am able to pass the test successfully on my pc. Any Thoughts?

**Comment by marco-c** [comment]
> @marco-c i've made all the minor changes you asked me to

It seems you haven't done https://github.com/marco-c/autowebcompat/pull/63#issuecomment-366649639 and https://github.com/marco-c/autowebcompat/pull/63#discussion_r169072493.

> i have removed the balance test and will open a pr for the same, because balance seems to be interacting with Couples-generator but we aren't writing tests for functions which need the data directory.

The balance function is generic, it doesn't require CouplesGenerator to work. Anyway yes, please open an issue to add a test for this function and an issue for each function in utils.py that isn't tested yet.

> One last thing is the error in the Travis-cl build it is giving an image not found error with the Load_image test, while I am able to pass the test successfully on my pc. 

Doing https://github.com/marco-c/autowebcompat/pull/63#discussion_r169072493 will fix the problem.

**Comment by Shashi456** [comment]
@marco-c i did modify both the load image file with the parent directory and try it with both temporary and an actual directory.  I will look further into, thankyou.

**Comment by marco-c** [comment]
> @marco-c i did modify both the load image file with the parent directory and try it with both temporary and an actual directory. I will look further into, thankyou.

Also don't download the image from the network, it's not great for unit tests (what if the image is removed from the website? The tests will start failing). Create a fake image using PIL.

File: tests/test_utils.py [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+    utils.mkdir(direc_path)
+    assert(os.path.isdir(direc_path))
+    utils.mkdir(direc_path)
+
+
+def test_load_image():
+    os.mkdir('data_resized')
+    fake_img = np.random.rand(30, 30, 3) * 255
+    img = Image.fromarray(fake_img.astype('uint8')).convert('RGBA')
+    img.save(".\data_resized\Image.jpg", "JPEG")
```
Comment (marco-c) [inline tests/test_utils.py]: Don't specify the format, it's automatically detected from the file extension.

**Comment by Shashi456** [comment]
@marco-c  I'll open the balance PR , as soon as this is merged. Everything works and i've done all changes needed. Thankyou

**Comment by Shashi456** [comment]
@marco-c i will open the balance PR as soon as this is merged, i have done all changes  you've asked me for . Thankyou

**Comment by marco-c** [comment]
> @marco-c i will open the balance PR as soon as this is merged, i have done all changes you've asked me for . Thankyou

I still don't see https://github.com/marco-c/autowebcompat/pull/63#issuecomment-366649639 :smile:

Also, in `test_load_image` you should use a temporary directory and not manually create and delete `data_resized`, which would mess the working directory of anyone which runs the tests.

File: tests/test_utils.py [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+    direc_path = os.path.join(d.name + "\\test")
+    utils.mkdir(direc_path)
+    assert(os.path.isdir(direc_path))
+    utils.mkdir(direc_path)
+
+
+def test_load_image():
+    os.mkdir('data_resized')
+    fake_img = np.random.rand(30, 30, 3) * 255
+    img = Image.fromarray(fake_img.astype('uint8')).convert('RGBA')
```
Comment (marco-c) [inline tests/test_utils.py]: Use `Image.new` instead of creating a numpy array and then convering it into an image.

File: tests/test_utils.py [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+    labels = utils.read_labels(file_name='labels.csv')
+    assert(isinstance(labels, dict))
+
+
+def test_write_labels():
+    label = {1: 1, 2: 2}
+    d = TemporaryDirectory()
+    file_path = os.path.join(d.name + "\\test.csv")
+    utils.write_labels(label, file_name=file_path)
+    assert(os.path.exists(file_path))
```
Comment (marco-c) [inline tests/test_utils.py]: Remember to file an issue about adding a test that writes labels and then reads them back.

File: tests/test_utils.py [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+
+
+def test_get_bugs():
+    bugs = utils.get_bugs()
+    assert(isinstance(bugs, list))
+
+
+def test_mkdir():
+    d = TemporaryDirectory()
+    direc_path = os.path.join(d.name + "\\test")
```
Comment (marco-c) [inline tests/test_utils.py]: You're calling `os.path.join` with a single string, which is basically useless.
Either do `os.path.join(d.name, 'test')` or `d.name + '\test'`.

File: tests/test_utils.py [hunk 18]
```diff
@@ -0,0 +1,42 @@
+import os
+import numpy as np
+from tempfile import TemporaryDirectory
+from autowebcompat import utils
+from PIL import Image
```
Comment (marco-c) [inline tests/test_utils.py]: Sort the imports in this way:
1) Built-in modules;
2) Modules from third-party libraries;
3) Our modules.

**Comment by Shashi456** [comment]
@marco-c do you necessarily want the new pytest feature and not my hack around it ?

File: .travis.yml [hunk 19]
```diff
@@ -6,3 +6,4 @@ install:
   - pip install -r test-requirements.txt
 script:
   - flake8 .
+  - pytest ./tests/test_utils.py
```
Comment (marco-c) [inline .travis.yml]: This is going to execute `test_utils.py` only, please make it run any test that is in the `tests` directory, so we don't have to update this line the next time we add a test.

**Comment by marco-c** [comment]
> @marco-c do you necessarily want the new pytest feature and not the Temporary Directory () way?

The pytest feature is nicer, so I would prefer it. If you can't, please file an issue for someone else to do it.

**Comment by Shashi456** [comment]
Would you want the PR's to be referenced to this?

**Comment by marco-c** [comment]
> Would you want the PR's to be referenced to this? @marco-c

Do you mean the issues? Yes

File: tests/test_utils.py [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+def test_load_image(tmpdir):
+    d = TemporaryDirectory()
+    img = Image.new("RGB", (30, 30))
+    img.save(d.name + "/Image.jpg")
+    img = utils.load_image("Image.jpg", d.name)
+    assert(isinstance(img, np.ndarray))
+
+
+def test_read_labels():
+    labels = utils.read_labels(file_name='labels.csv')
```
Comment (marco-c) [inline tests/test_utils.py]: Create a fake labels.csv file in a temporary directory, so you can assert more things about the result.

Comment (marco-c) [reply tests/test_utils.py]: Actually, don't do it, we'll do it as part of the other test that writes and reads labels.

File: tests/test_utils.py [hunk 21]
```diff
@@ -0,0 +1,39 @@
...
+    assert(os.path.isdir(direc_path))
+    utils.mkdir(direc_path)
+
+
+def test_load_image(tmpdir):
+    d = TemporaryDirectory()
+    img = Image.new("RGB", (30, 30))
+    img.save(d.name + "/Image.jpg")
+    img = utils.load_image("Image.jpg", d.name)
+    assert(isinstance(img, np.ndarray))
```
Comment (marco-c) [inline tests/test_utils.py]: You can assert the dimensions of the numpy array too.

File: .travis.yml (hunk 19 above)
Comment (marco-c) [reply .travis.yml]: Don't forget about this, @Shashi456
